org.eclipse.scanning.event/src/org/eclipse/scanning/event/ConsumerImpl.java

```diff
@@ -416,86 +451,141 @@ private void terminateProcess(IConsumerProcess<U> process) throws EventException
+		// NOTE: Not all consumers check the submit queue and
+		// pause before they start.
+        checkStartPaused();
+		
+		// It is possible to call start() and then awaitStart().
+		if (latchStart!=null) latchStart.countDown();
+	}
 
-        		
-        		if (!isDurable()) break;
-        		        		
-       			try {
-       				if (Thread.interrupted()) break;
-					Thread.sleep(Constants.getNotificationFrequency());
-				} catch (InterruptedException e) {
-					throw new EventException("The consumer was unable to wait!", e);
-				}
-       			
-       			waitTime+=Constants.getNotificationFrequency();
-    			checkTime(waitTime); 
-    			
-        		logger.warn(getName()+" ActiveMQ connection to "+uri+" lost.");
-        		logger.warn("We will check every 2 seconds for 24 hours, until it comes back.");
-
-        		continue;
-        	}
+	private boolean consume() throws Exception {
+		
+		checkPaused(); // blocks until not paused.
+		if (!isActive()) return false; // Might have pasued for a long time.
+		
+		// Consumes messages from the queue.
+    	Message m = getMessage(uri, getSubmitQueueName());
+        if (m!=null) {
+        	waitTime = 0; // We got a message
+        	
+        	// TODO FIXME Check if we have the max number of processes
+        	// exceeded and wait until we don't...
+        	
+        	TextMessage t = (TextMessage)m;
+        	
+        	final String json  = t.getText();
+        	
+			@SuppressWarnings("unchecked")
+			final U bean   = (U) service.unmarshal(json, getBeanClass());
+            
+        	executeBean(bean);   
+        }
+        return true;
+	}
+
+	private boolean processException(Throwable ne) throws EventException {
+
+		if (ne instanceof EventException || ne instanceof InterruptedException) {
+			if (Thread.interrupted()) return false;
+			logger.error("Cannot consume message ", ne);
+			if (isDurable()) {
+				return true;
+			}
+			return false;
+		}
+
+		logger.debug("Error in consumer!", ne);
+		if (ne.getClass().getSimpleName().contains("Json")) {
+			logger.error("Fatal except deserializing object!", ne);
+			return true;
+		}
+		if (ne.getClass().getSimpleName().endsWith("UnrecognizedPropertyException")) {
+			logger.error("Cannot deserialize bean!", ne);
+			return true;
+		}
 
+		if (ne.getClass().getSimpleName().endsWith("ClassCastException")) {
+			logger.error("Problem with serialization?", ne);
 		}
+
+
+		if (!isDurable()) return false;
+
+		try {
+			if (Thread.interrupted()) return false;
+			Thread.sleep(Constants.getNotificationFrequency());
+		} catch (InterruptedException e) {
+			throw new EventException("The consumer was unable to wait!", e);
+		}
+
+		waitTime+=Constants.getNotificationFrequency();
+		checkTime(waitTime); 
+
+		logger.warn(getName()+" ActiveMQ connection to "+uri+" lost.");
+		logger.warn("We will check every 2 seconds for 24 hours, until it comes back.");
+
+		return true;
 	}
-	
+
+	private void checkStartPaused() throws EventException {
+		
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1155)